go/heif/heif.go

```diff
@@ -602,43 +694,259 @@ func (img *Image) GetImage() (image.Image, error) {
+		Plane:   C.GoBytes(pointer, size),
+		Stride:  int(stride),
+		Height:  int(height),
+		Pointer: pointer,
+		image:   img,
 	}
 	return access, nil
 }
 
+func (img *Image) NewPlane(channel Channel, width, height, depth int) (*ImageAccess, error) {
+	err := C.heif_image_add_plane(img.image, uint32(channel), C.int(width), C.int(height), C.int(depth))
+	if err := convertHeifError(err); err != nil {
+		return nil, err
+	}
+	return img.GetPlane(channel)
+}
+
 func (img *Image) ScaleImage(width int, height int) (*Image, error) {
 	var scaled_image Image
 	err := C.heif_image_scale_image(img.image, &scaled_image.image, C.int(width), C.int(height), nil)
 	if err := convertHeifError(err); err != nil {
 		return nil, err
 	}
 
-	runtime.SetFinalizer(&scaled_image, free_heif_image)
+	runtime.SetFinalizer(&scaled_image, freeHeifImage)
 	return &scaled_image, nil
 }
 
+// --- High-level encoding API.
+
+type EncodingOptions struct {
+	options *C.struct_heif_encoding_options
+}
+
+func NewEncodingOptions() (*EncodingOptions, error) {
+	options := &EncodingOptions{
+		options: C.heif_encoding_options_alloc(),
+	}
+	if options.options == nil {
+		return nil, fmt.Errorf("Could not allocate encoding options")
+	}
+
+	runtime.SetFinalizer(options, freeHeifEncodingOptions)
+	return options, nil
+}
+
+func freeHeifEncodingOptions(options *EncodingOptions) {
+	C.heif_encoding_options_free(options.options)
+	options.options = nil
+	runtime.SetFinalizer(options, nil)
+}
+
+func hasAlpha(i image.Image) bool {
+	rect := i.Bounds()
+	for y := rect.Min.Y; y < rect.Max.Y; y++ {
+		for x := rect.Min.X; x < rect.Max.X; x++ {
+			if _, _, _, c := i.At(x, y).RGBA(); c == 0xffff {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func imageFromNRGBA(i *image.NRGBA) (*Image, error) {
+	max := i.Bounds().Max
+	w, h := max.X, max.Y
+
+	var cm Chroma
+	var depth int
+	var stride int
+	if hasAlpha(i) {
+		cm = ChromaInterleavedRGBA
+		depth = 32
+		stride = w * 4
+	} else {
+		cm = ChromaInterleavedRGB
+		depth = 64
+		stride = w * 3
+	}
+
+	out, err := NewImage(w, h, ColorspaceRGB, cm)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create image: %v", err)
+	}
+
+	p, err := out.NewPlane(ChannelInterleaved, w, h, depth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add plane: %v", err)
+	}
+	p.SetData([]byte(i.Pix), stride)
+
+	return out, nil
+}
+
+func imageFromGray(i *image.Gray) (*Image, error) {
+	max := i.Bounds().Max
+	w, h := max.X, max.Y
+
+	out, err := NewImage(w, h, ColorspaceYCbCr, ChromaMonochrome)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create image: %v", err)
+	}
+
+	const depth = 8
+	pY, err := out.NewPlane(ChannelY, w, h, depth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add Y plane: %v", err)
+	}
+	pY.SetData([]byte(i.Pix), i.Stride)
+
+	return out, nil
+}
+
+func imageFromYCbCr(i *image.YCbCr) (*Image, error) {
+	max := i.Bounds().Max
+	w, h := max.X, max.Y
+
+	out, err := NewImage(w, h, ColorspaceYCbCr, Chroma420)
```

Comment:
You need to check the `SubsampleRatio` of the `image.YCbCr`, it could also be `4:2:2` or `4:4:4`.

Reply:
Like that?

```go
var chroma Chroma
switch i.SubsampleRatio {
case image.YCbCrSubsampleRatio444:
	chroma = Chroma444
case image.YCbCrSubsampleRatio422:
	chroma = Chroma422
case image.YCbCrSubsampleRatio420:
	fallthrough
default:
	chroma = Chroma420
}
```